src/main/java/io/nats/client/impl/MessageQueue.java

```diff
@@ -27,39 +28,30 @@
     private final static int RUNNING = 1;
     private final static int DRAINING = 2;
 
-    public static final int MAX_SPINS = 200; // Default max spins
-    public static final int SPIN_WAIT = 50;
-    public static final int MAX_SPIN_TIME = SPIN_WAIT * MAX_SPINS;
-
     private final AtomicLong length;
     private final AtomicLong sizeInBytes;
     private final AtomicInteger running;
     private final boolean singleThreadedReader;
-    private final int maxSpins;
-    private final ConcurrentLinkedQueue<NatsMessage> queue;
-    private final ConcurrentLinkedQueue<Thread> waiters;
+    private final LinkedBlockingQueue<NatsMessage> queue;
     private final Lock filterLock;
+    private final NatsMessage poisonPill;
 
-    MessageQueue(boolean singleReaderMode) {
-        this.queue = new ConcurrentLinkedQueue<>();
+    MessageQueue(boolean singleReaderMode, int publishHighwaterMark) {
+        this.queue = publishHighwaterMark > 0 ? new LinkedBlockingQueue<NatsMessage>(publishHighwaterMark) : new LinkedBlockingQueue<NatsMessage>();

```

Comment:
What does default do with no publishHighwaterMark?

Reply:
when set to 0 we have an unbounded queue

Reply:
That ever good?

Reply:
only the reader uses 0 for subscribers, so it is checked a different way.